Accept a closure for `with` in lieu of a path for fields

TedDriggs opened the pull request:
Fixes #309

This has some issues, so it's not yet ready to merge:

1. It doesn't work for all the places where `with` is used.
2. I'm not thrilled with the introduction of new locals to support this, and am wondering if there's a way to write it that better encapsulates the desired behavior.

Veetaha commented:
If you'd like to avoid using a local variable there are two ways to do that:

The naive way us to just cast the expression with `as`: `#closure as fn(...)`. Although the `as` operator is a bit more powerful for this task, because it's meant for potential type coersions, while in this case we just need a type hint.

Another way to do that is with `::core::convert::identity::<fn(...) -> _>(#closure)` which provides a hint for the expression

TedDriggs commented:
Using `identity` seems to work well. It also solves one of the main barriers to allowing arbitrary expressions in this position, though I'm still hesitant to allow arbitrary expressions until I can test the isolation and the quality of the error messages when the user does something wrong.

TedDriggs commented:
@Veetaha I've updated this using `core::convert::identity`, added a compile-fail test to show what happens if you try to access locals, and renamed `darling_core::codegen::Field::with_path` to be the more descriptive `with_callable`.

Veetaha reviewed (approved):
LGTM, except docs should need a small update

TedDriggs commented:
@Veetaha ugh, one more question for you: Any ideas on how to fix the output of the compiletest so it will pass both locally and on CI? It seems the issue is the path on the last line of the error, potentially due to the use of `$RUST` vs a specified path?

Veetaha commented:
I haven't used `compiletest`. I assume it's an older equivalent of [`trybuild`](https://docs.rs/trybuild/latest/trybuild/), which I've been using for my tests. Maybe there it will work with `trybuild`?

The version of trybuild, that supports 1.56 MSRV is `1.0.89` (can be found in the [table here](https://lib.rs/crates/trybuild/versions)).

TedDriggs commented:
I'm using `trybuild` - the `compiletests` is just the test directory name, since I didn't want to confuse it by recycling the name.

These tests are run on 1.77.0; I'm wondering if the issue is the lack of an environment variable for the Rust directory in the local `compiletests.sh` file, since it looks like the CI version of the test is using `$RUST` where the local version used that absolute path.

TedDriggs commented:
Okay, it looks like `trybuild@1.0.89` is the sweet spot where builds on 1.56.0 still work, though why CI is ever trying to build with `trybuild` on 1.56.0 is unclear to me, as that dependency is only present for `cfg(compiletests)` and that's only used on 1.77.0.

Veetaha commented:
>  why CI is ever trying to build with trybuild on 1.56.0 is unclear to me, as that dependency is only present for cfg(compiletests) and that's only used on 1.77.0

Do you mean this error?:

> failed to select a version for the requirement

If yes, then that's expected because this happens not at the building stage, but when updating the crates.io index. `cargo` needs to make sure the dependency exists (even if its under a cfg), and create a single `Cargo.lock` file which lists the pinned dependency versions assuming all dependencies are used

TedDriggs commented:
Yes, that's the error I meant. Thanks for the explanation. For now, it seems that leaving the dependency at 1.0.89 is sufficient. If it ever breaks, that'll probably be a sign it's time to bump `darling`'s MSRV for its next release.

I've updated examples, tests, and CHANGELOG for this feature; I didn't see any places in the actual docs where a change was needed. I do want to look at expanding this to cover the other `with` scenarios, though I don't think that needs to be part of this PR.
